rfd/0234-in-band-mfa-ssh-sessions.md

```diff
@@ -0,0 +1,756 @@
+new SSH connection.
+
+#### Local Cluster Flow
+
+```mermaid
+---
+title: SSH MFA Connection Flow (Local Cluster)
+---
+sequenceDiagram
+  autoNumber
+
+  participant Client
+  participant Proxy as Proxy Service
+  participant MFA as MFA Service
+  participant SSH as SSH Service
+  participant Decision Service
+  participant Host as Target SSH Host
+  Client->>Proxy: Dial SSH
+
+  Proxy->>Decision Service: EvaluateSSHAccess
+  Decision Service-->>Proxy: Permit
+  Proxy->>SSH: Proxy SSH connection (stapled permit)
+  SSH->>SSH: Is MFA required?
+
+  alt MFA required
+    SSH->>Client: Keyboard-interactive (MFA required)
+    Client->>MFA: CreateChallenge (SIP)
+    MFA-->>MFA: Generate and store MFA challenge (SIP)
+    MFA-->>Client: MFA challenge
+
+    Client->>Client: Solve MFA challenge
+    Client->>MFA: ValidateChallenge (MFA response)
+    MFA-->>MFA: Validate MFA response
+    MFA-->>Client: Challenge validated
+
+    Client->>SSH: Keyboard-interactive (challenge name)
+    SSH->>MFA: GetValidatedChallenge (challenge name)
+    MFA-->>SSH: ValidatedChallengeResponse and SIP
+    SSH-->>SSH: Compute SIP
+    SSH-->>SSH: Verify SIP matches
+  end
+
+  SSH->>Host: Establish SSH connection
+  Host-->>SSH: SSH connection established
+  SSH-->>Proxy: SSH connection established
+  Proxy-->>Client: SSH session established
+```
+
+#### Leaf Cluster Flow
+
+```mermaid
+---
+title: SSH MFA Connection Flow (Leaf Cluster)
+---
+sequenceDiagram
+  autoNumber
+
+  participant Client
+  participant rMFA as Local MFA Service
+  participant Proxy as Leaf Proxy Service
+  participant Dec as Leaf Decision Service
+  participant lMFA as Leaf MFA Service
+  participant SSH as Leaf SSH Service
+
+  participant Host as Target SSH Host
+
+  Client->>Proxy: Dial SSH
+  Proxy->>Dec: EvaluateSSHAccess
+  Dec-->>Proxy: Permit
+  Proxy->>SSH: Proxy SSH connection (stapled permit)
+  SSH->>SSH: Is MFA required?
+
+  alt MFA required
+    SSH->>Client: Keyboard-interactive (MFA required)
+    Client->>rMFA: CreateChallenge (SIP)
+    rMFA-->>rMFA: Generate and store MFA challenge with SIP
+    rMFA-->>Client: MFA challenge
+
+    Client->>Client: Solve MFA challenge
+    Client->>rMFA: ValidateChallenge (MFA response)
+    rMFA-->>rMFA: Validate MFA response
+    rMFA-->>lMFA: ReplicateValidatedChallenge (SIP)
+    rMFA-->>Client: Challenge validated
+
+    Client->>SSH: Keyboard-interactive (challenge name)
+    SSH->>lMFA: GetValidatedChallenge (challenge name)
+    lMFA-->>SSH: ValidatedChallengeResponse and SIP
+    SSH-->>SSH: Compute SIP
+    SSH-->>SSH: Verify SIP matches
+  end
+
+  SSH->>Host: Establish SSH connection
+  Host-->>SSH: SSH connection established
+  SSH-->>Proxy: SSH connection established
+  Proxy-->>Client: SSH session established
+```
+
+### Security
+
+In addition to the risks raised in [Access Control Decision API (RFD
+0024e)](https://github.com/gravitational/Teleport.e/blob/master/rfd/0024e-access-control-decision-api.md), there are a
```

Comment:
> In addition to the risks raised in Access Control Decision API

Are the risks you mentioned because this relies on the Decision API, or is there some particular risk that applies here? I think it pays to be explicit in what you meant here.